core/sail/api/src/main/java/org/eclipse/rdf4j/sail/helpers/SailConnectionWrapper.java

```diff
@@ -206,6 +207,11 @@ public Explanation explain(Explanation.Level level, TupleExpr tupleExpr, Dataset
 		return wrappedCon.explain(level, tupleExpr, dataset, bindings, includeInferred, timeoutSeconds);
 	}
 
+	@Override
+	public void begin(TransactionSetting... settings) {
+		wrappedCon.begin(settings);
+	}
+
 	@Override
 	public void begin() throws SailException {
 		wrappedCon.begin();
```

Comment:
@jeenbroekstra 

The whole wrapped connection approach here makes it hard for me to add support for transaction settings. The problem is that code higher up will often override the `begin()` and `begin(IsolationLevel level)` methods to hook into the transaction handling approach. With the new `begin(TransactionSetting[] settings)` simply calling `wrapped.begin(settings)` the code higher up will have to override the new begin method in order to fully hook into the transaction flow.

I'm not sure right now how to fix this. I was initially hoping to hook in like I did in the AbstractSailConnection.

Reply:
Maybe I want to try out just extending the current `begin(IsolationLevel level)` method with the transaction settings vararg, and in the next major release remove that in favour of just a simpler `begin(TransactionSetting... settings)`.

Reply:
I actually tried this out now.

```
void begin(IsolationLevel level, TransactionSetting ... settings)
```

And it's not backwards compatible with any of the code that implements the current `begin` interface methods. I'm getting a bunch of compile time errors saying that various classes have `@Override` annotation for methods that are not inherited.

Also found another issue with introducing a new begin method that only takes a TransactionSetting vararg.

<img width="821" alt="Screenshot 2020-07-04 at 17 54 29" src="https://user-images.githubusercontent.com/797185/86516273-738b6400-be1f-11ea-81e9-b2c50976cbf0.png">

This isn't exactly good code though, since it making the isolation level nullable means that all code actually needs to check for null. Instead of just having is non-null and instead passing in a default isolation level if none was defined.

Reply:
You can avoid this by explicitlying cast the null argument to a TransactionSetting for this to work I think:

    this.begin((TransactionSetting)null);

That way the vararg itself is never null, it just receives an array with a single element, and that element is null.

Reply:
You can avoid this by explicitlying cast the null argument to a TransactionSetting for this to work I think:

    this.begin((TransactionSetting)null);

That way the vararg itself is never null, it just receives an array with a single element, and that element is null.

Reply:
> Maybe I want to try out just extending the current `begin(IsolationLevel level)` method with the transaction settings vararg, and in the next major release remove that in favour of just a simpler `begin(TransactionSetting... settings)`.

That sound like a good idea to me.

Reply:
I tried adding a vararg, but it's not source level compatible.

Reply:
> @jeenbroekstra I tried adding a vararg, but it's not source level backwards compatible.

I'm sorry I'm not sure what exact variant we're talking about here.

Reply:
begin(IsolationLevel level, TransactionSettings ... settings)

If I add the vararg to the current begin method like above it’s not source level compatible.

Reply:
Not even if you also keep 

    begin(IsolationLevel level) 

and just redirect it to:

    begin(level, (TransactionSettings)null);

?

Reply:
Could redirect it in the sail wrapper, but it’ll introduce its own problems.